portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import com.google.gson.Gson;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    UserService userService = UserServiceFactory.getUserService();
+
+    // If user is not logged in, show a login form (could also redirect to a login page)
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL("/login");
+      out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
```

Comment:
It's not ideal to be generating client-side HTML in a string on the server.  Instead, could you just return some JSON that has a boolean for if the user is logged in, and any other data you need?  That way the client-side code can respond however it needs to.

Same comment throughout.  It's much better to separate client-side and server-side code.  The server should just be passing raw data to the client, who decides on formatting and layout.  It also makes the code much more reusable.  For example, what if you want to log in from one of the pages the other interns are making?  The servlet shouldn't need to know what page you're on